test/e2e/apimachinery/garbage_collector.go

```diff
@@ -719,4 +725,105 @@ var _ = SIGDescribe("Garbage collector", func() {
+				framework.Failf("failed to delete CustomResourceDefinition: %v", err)
+			}
+		}()
+		client, err := apiextensionstestserver.CreateNewCustomResourceDefinition(definition, apiExtensionClient, f.ClientPool)
+		if err != nil {
+			framework.Failf("failed to create CustomResourceDefinition: %v", err)
+		}
+
+		// Get a client for the custom resource.
+		resourceClient := client.Resource(&metav1.APIResource{
+			Name:       definition.Spec.Names.Plural,
+			Namespaced: false,
+		}, api.NamespaceNone)
+		apiVersion := definition.Spec.Group + "/" + definition.Spec.Version
+
+		// Create a custom parent resource.
+		parentName := names.SimpleNameGenerator.GenerateName("parent")
+		parent := &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": apiVersion,
+				"kind":       definition.Spec.Names.Kind,
+				"metadata": map[string]interface{}{
+					"name": parentName,
+				},
+			},
+		}
+		persistedParent, err := resourceClient.Create(parent)
+		if err != nil {
+			framework.Failf("failed to create parent resource %q: %v", parentName, err)
+		}
+		framework.Logf("created parent resource %q", parentName)
+
+		// Create a custom child resource.
+		childName := names.SimpleNameGenerator.GenerateName("child")
+		child := &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": apiVersion,
+				"kind":       definition.Spec.Names.Kind,
+				"metadata": map[string]interface{}{
+					"name": childName,
+					"ownerReferences": []map[string]string{
+						{
+							"uid":        string(persistedParent.GetUID()),
+							"apiVersion": apiVersion,
+							"kind":       definition.Spec.Names.Kind,
+							"name":       parentName,
+						},
+					},
+				},
+			},
+		}
+		persistedChild, err := resourceClient.Create(child)
+		if err != nil {
+			framework.Failf("failed to create child resource %q: %v", childName, err)
+		}
+		framework.Logf("created child resource %q", childName)
+
+		// Delete the parent.
+		background := metav1.DeletePropagationBackground
+		err = resourceClient.Delete(parentName, &metav1.DeleteOptions{PropagationPolicy: &background})
+		if err != nil {
+			framework.Failf("failed to delete parent resource %q: %v", parentName, err)
+		}
+
+		// Ensure the child is deleted.
+		if err := wait.Poll(5*time.Second, 60*time.Second, func() (bool, error) {
+			_, err := resourceClient.Get(childName, metav1.GetOptions{})
+			return errors.IsNotFound(err), nil
+		}); err != nil {
+			framework.Logf("parent: %#v", persistedParent)
+			framework.Logf("child: %#v", persistedChild)
+			framework.Failf("failed waiting for child resource %q to be deleted", childName)
+		}
+
+		// Ensure the parent is deleted.
+		if err := wait.Poll(5*time.Second, 60*time.Second, func() (bool, error) {
+			_, err := resourceClient.Get(parentName, metav1.GetOptions{})
+			return errors.IsNotFound(err), nil
+		}); err != nil {
+			framework.Failf("failed waiting for parent resource %q to be deleted", parentName)
+		}
```

Comment:
After the deletion is completed, can we delete the CRD definition, then recreate the CRD definition, then create the owner and dependents, then see if cascading deletion still works? I want to verify that the GC can handle repeated creation/deletion of CRD definition.

Can we add a few more test scenarios:
1. owner is a CRD but the dependents are core kubernetes API (like pod), and vice versa.
2. create a CRD, make a pod depend on the CRD, then delete the CRD definition, sleep 30 seconds to make sure the monitors are resynced, verify that the pod is not deleted; then recreate the CRD definition, sleep another 30 seconds, delete the CRD object, verify if the pod is deleted. 
3. Create a CRD object, make it depend on a pod, and set the ownerReference.BlockOwnerDeletion=true; delete the CRD definitions, sleep 30 seconds, then delete the pod with policy=Foreground, verify that the pod is deleted (i.e., not blocked on the disappeared CRD definition).

Maybe not all these cases make sense. Let's brainstorm on if there are other corner cases.